google-cloud-spanner/lib/google/cloud/spanner/interval.rb

```diff
@@ -0,0 +1,270 @@
+      class Interval
+        NANOSECONDS_IN_A_SECOND = 1_000_000_000
+        NANOSECONDS_IN_A_MINUTE = NANOSECONDS_IN_A_SECOND * 60
+        NANOSECONDS_IN_AN_HOUR = NANOSECONDS_IN_A_MINUTE * 60
+        NANOSECONDS_IN_A_MILLISECOND = 1_000_000
+        NANOSECONDS_IN_A_MICROSECOND = 1_000
+        MAX_MONTHS = 120_000
+        MIN_MONTHS = -MAX_MONTHS
+        MAX_DAYS = 3_660_000
+        MIN_DAYS = -MAX_DAYS
+        MAX_NANOSECONDS = 316_224_000_000_000_000_000
+        MIN_NANOSECONDS = -316_224_000_000_000_000_000
+
+        private_constant :NANOSECONDS_IN_A_SECOND, :NANOSECONDS_IN_A_MINUTE, :NANOSECONDS_IN_AN_HOUR,
+                         :NANOSECONDS_IN_A_MILLISECOND, :NANOSECONDS_IN_A_MICROSECOND, :MAX_MONTHS,
+                         :MIN_MONTHS, :MAX_DAYS, :MIN_DAYS, :MAX_NANOSECONDS, :MIN_NANOSECONDS
+
+        class << self
+          # Parses an ISO8601 string and returns an Interval instance.
+          # The accepted format for the ISO8601 standard is:
+          # `P[n]Y[n]M[n]DT[n]H[n]M[n[.fraction]]S`
+          # where `n` represents an integer number.
+          #
+          # @param [String] An ISO8601 formatted string.
+          # @return [Google::Cloud::Spanner::Interval]
+          #
+          # @example
+          #   require "google/cloud/spanner"
+          #
+          #   iso_8601_string = "P1Y2M3DT4H5M6S"
+          #   interval = Google::Cloud::Spanner::Interval::parse iso_8601_string
+          #
+          #   puts interval # "P1Y2M3DT4H5M6S"
+          def parse interval_string
+            pattern = /^P(?!$)((?<years>-?\d+)Y)?((?<months>-?\d+)M)?((?<days>-?\d+)D)?(T(?!$)((?<hours>-?\d+)H)?((?<minutes>-?\d+)M)?((?<seconds>-?(?!S)\d*([\.,]\d{1,9})?)S)?)?$/
+            interval_months = 0
+            interval_days = 0
+            interval_nanoseconds = 0
+
+            matches = interval_string.match pattern
+
+            if matches.nil?
+              raise ArgumentError, "The provided string does not follow ISO8601 standard."
+            end
+
+            if matches.captures.empty?
+              raise ArgumentError, "The provided string does not follow ISO8601 standard."
+            end
+
+            if matches[:years]
+              interval_months += matches[:years].to_i * 12
+            end
+
+            if matches[:months]
+              interval_months += matches[:months].to_i
+            end
+
+            if matches[:days]
+              interval_days = matches[:days].to_i
+            end
+
+            if matches[:hours]
+              interval_nanoseconds += matches[:hours].to_i * NANOSECONDS_IN_AN_HOUR
+            end
+
+            if matches[:minutes]
+              interval_nanoseconds += matches[:minutes].to_i * NANOSECONDS_IN_A_MINUTE
+            end
+
+            # Only seconds can be fractional. Both period and comma are valid inputs.
+            if matches[:seconds]
+              interval_nanoseconds += matches[:seconds].gsub(",", ".").to_f * NANOSECONDS_IN_A_SECOND
```

Comment:
This will result in a float type. We should cast back `to_i` after multiplying by `NANOSECONDS_IN_A_SECOND`.